### ISSUE TITLE
Automatically Populate Super.Human.Installer Vagrant Instances

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#1217]
Currently we need to manually setup new Vagrant instances in Moonshine:
1. Open Moonshine > Settings > Vagrant
2. Add a new Linked Instance
    - Label:  The most straightforward option is to match the server.
    - Base URL;  This URL is not provided currently on the SHI side, but the syntax is `http://restapi.%server_hostname%:8080`.  Note that `restapi.%server_hostname%` can be replaced with the IP address, but the user is encouraged to update the hosts file anyway.
    - Local Path is optional

We would like to have Moonshine automatically detect and configure the Linked Instances.  The proposed logic is:
1. The new logic will be triggered on startup, or when SHI sends an application message to Moonshine when the server finishes provisioning (similar to MSDKI)
2. Moonshine will check the SHI and production and development configuration files and find all configured servers.
3. Moonshine will then add Linked Instance entries for each server, if they don't exist already.  Use the server name for the Label and Base URL, as described above.  To check if a server exists already, compare the server name to the instance names.
4. Prompt the user to show the new server.  This should have similar information to what appears in the Linked Vagrant Instance form (or we can use the form itself).  Note that multiple instance could be installed at the same time.

IMAGE

We can find the configuration files at:

    Prod: ~/Library/Application Support/SuperHumanInstaller/.shi-config
    Dev:  ~/Library/Application Support/SuperHumanInstallerDev/.shi-config

```
 % cat .shi-config 
{
	"servers": [
		{
			"provisioner": {
                                ... We could potentially check the provisioner for known supported types, but I think it is more flexible rely on the configured RESTAPI capabilities ...
				"type": "demo-tasks",
				"version": 1020
			},
			"roles": [
				... these could be relevant if we want to limit which servers are automatically configured.  For now, we only have one type of server, though...
			],
			"server_organization": "DEMOPROD",
			"server_hostname": "domino-1.demoprod.com",
			"type": "domino",
			"server_id": 1827,
                        ...
		},
		... more servers ...
	],

```

So, we should use `server_hostname` for the Label and Base URL.

It should be easy to populate the Local Path as well.  Collect the `server_id` and `provisioner.type` from the JSON files, and append it to these paths to get the full path:

    Prod: ~/Library/Application Support/SuperHumanInstaller/servers/%provisioner.type%/%server_id%
    Dev:  ~/Library/Application Support/SuperHumanInstallerDev/servers/%provisioner.type%/%server_id%

Limitations:
- The server name won't work for the URL until the user updates the hosts file.  Unfortunately, this means that the server configuration will almost always be invalid when the message to Moonshine triggers.  The entry can be validated once the hosts file is updated.
- MacOS may block access to the `Application Support` files for other applications.  If needed, we could update SHI to write configuration to a directory in ~/Downloads

UPDATE:  There is a special case for `server_hostname` where the user enters something like "demo" instead of a full FQDN.  In this case, SHI appends `.%server_organization%` to get the full servername, but this is not stored in `server_hostname`.  For now, please add the same logic on the Moonshine side.  I updated the JSON above to show an example of `server_organization`

UPDATE:
Branch:  `issue_1217_read_shi_config`
Builds:
- MacOS:  https://static.moonshine-ide.com/downloads/releases/dev/macos/branches/issue_1217_read_shi_config/
- Windows:  TODO

**Comment by rat-moonshine** [comment]
I don't see I have any `.shi_config` in the folder location. Please consider to provide one for a test. Meanwhile I shall put a demo .json that matches the aforementioned content.

**Comment by rat-moonshine** [comment]
Small correction: It seems the [mentioned](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1217#issue-1843895404) SHI application storage path on macOS is:

`~/Library/Application Support/Prominic.NET/SuperHumanInstaller/.shi-config`

I gave some tests to access the file/location in a packaged installation. Following were my tests results on accessing the file:
Path | Accessible
---------|------------
`/Users/$user/Library/Application Support/Prominic.NET/SuperHumanInstallerDev/.shi_config` | `File` API sees does not exists
`/Users/$user/Library/Application Support/Prominic.NET/SuperHumanInstallerDev/.shi_config.json` | `File` API able to access/read

**Comment by rat-moonshine** [comment]
This the code-trail for activating Moonshine-IDE from MSDKI - https://github.com/Moonshine-IDE/Moonshine-SDK-Installer/blob/master/MoonshineSDKInstaller/src/actionScripts/utils/MoonshineNotifier.as#L64. We requires to call this after necessary file being generated.

**Comment by rat-moonshine** [comment]
Okay. I think I have done with most of the functionalities that needs at Moonshine-IDE side. Following is a brief on implementation updates:

- Moonshine reads from `.shi_config` file for both production and development SHI installations
- Moonshine reads and parse instances from the configs during startup and generates the Vagrant Instances list
- Moonshine _can_ react on messaging from other applications, i.e. from-SHI, and re-read the configs and update the list (how to trigger Moonshine in this case I have already noted [here](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1217#issuecomment-1673588085))
- Moonshine auto-opens the `Settings > Vagrant` section in above messaging case
- Moonshine then open the first or only instance in view/edit popup from the parsed list

**Comment by JoelProminic** [comment]
I do want better feedback for the user when new Vagrant Instances are added.  With the current logic, this will only trigger when a message is sent from SHI, and only for one instance. 

For a simple initial solution, please add a message to the console for each added (not already existing) Vagrant Instance:

> Added a new Vagrant Instance (%Label%).  Check Settings > Vagrant for more details. 

In the long term, we may want to add a small prompt for this, or otherwise make it easier for the user to see the updates.

**Comment by JoelProminic** [comment]
I downloaded the latest branch build (508) after today's meeting and did some tests, but I ran into some problems:

Deleting Vagrant instances doesn't seem to work.  I have tried:
1. Open Settings > Vagrant
2. Change the label for one of the imported custom instances and save
3. Reopen Settings > Vagrant 
4. Delete all Vagrant Instances and save
5. Reopen Settings > Vagrant.  The instances are gone
6. Restart Moonshine
7. Reopen Settings > Vagrant.  The SHI instances were recreated, but I also see an instance with the modified Label

Similarly, I'm seeing that the "demo" instance that was imported with the previous test build is restored, and Moonshine doesn't import a new instance as "demo.startcloud.com" (I see that the URL is wrong, so this should be the old entry).  When checking for existing entries on the "demo" case, you should check for the full computed name instead.

**Comment by rat-moonshine** [comment]
I confirm the editing problem. Please, update with a new Moonshine build, and give a fresh test to ensure no-conflict because of old-data structure. 

- Items are now properly saving upon editing
- Moonshine now keeps way to compare between original and modified entry with SHI configuration
- Moonshine now compares loaded entries using calculated title

However, once deleted any SHI entry shall re-populates again when Moonshine restarts. Moonshine reads the configuration during startup, and currently it doesn't has a mechanism to keep track of once deleted SHI entry. To do that, we might require to maintain another cache/list contains once deleted SHI entry references. I'm not sure if we maintain that much of overhead in this case, but if we needs it that way - then I shall have this implemented. 

Please, check.

**Comment by JoelProminic** [comment]
The current logic is looking good, so please merge it so that @JustinProminic can test this more easily.

I also want to do some quick tests for this to confirm it does not return errors when SHI and/or SHI dev are not installed

**Comment by rat-moonshine** [comment]
It is now merged into `master`.

**Comment by JoelProminic** [comment]
I can see this working with both SHI and SHI Development with the master build.

**Comment by JoelProminic** [comment]
I've seen a couple cases in the past couple weeks where my Vagrant Instance list cleared except for a `http://restapi...com:8080` entry:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/50b63c9a-9752-4743-b71d-b69806892958)

I haven't yet figured out what event triggers this.   The empty entry looks like it was an invalid entry generated by the logic in this issue, so I'm reopening this issue for a quick investigation (@rat-moonshine).   If we can't reproduce this, or find it is an unrelated problem, we can move this to a separate issue.  If someone else has seen this problem, please add a note to the issue.

**Comment by JoelProminic** [comment]
@rat-moonshine do you have any immediate ideas for what could cause this?

**Comment by rat-moonshine** [comment]
Sorry, I haven't received the earlier comment/notification, for some reason.

It looks like partly-populated data; and I looked into the code and here's where such population can happened - during reading from SHI configuration to populate the list:
https://github.com/Moonshine-IDE/Moonshine-IDE/blob/master/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/vagrant/utils/VagrantUtil.as#L192

These population codes are old, and if this suddenly starts breaking for you, you might want to check the file as pointed [here](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/master/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/vagrant/utils/VagrantUtil.as#L150).

**Comment by JoelProminic** [comment]
Here is a [better link](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/7e96614e6c80b34986c90b5b1536794c327aee63/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/vagrant/utils/VagrantUtil.as#L72-L79) for the file paths.

Checking the files, I see I only have an incomplete server configuration defined in Super.Human.Installer.Dev, so this would explain the incomplete entry.  I thought we had added some logic to only include provisioned servers, but I don't see a flag for this in the JSON:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/6a7fa9f5-ca08-48a9-a0bf-d8f054ad3b92)

However, this doesn't explain why my manual Vagrant instances were cleared, and this is the bigger problem for me in this case.

**Comment by rat-moonshine** [comment]
It appears [this](https://github.com/Moonshine-IDE/Moonshine-IDE/commit/f0ed559f27b524690db6b0b8e096b0b456cd9dbf#diff-7e94bf51f3cca2e6b1a3867641e0ab4e708f3d081263a6019176cd428633ec7bL132) condition was blocking re-population of previously-saved/manually-entered items to list. 

`vagrantInstance.server` is suppose to generates when read from .shi file - and NOT generates in a manually-entered item. So, when Moonshine starts next time, and populates the list by reading from saved cookies - the condition was trying to ensure that `vagrantInstance.server` presents and `vagrant.server.hostName` is not _null_ - these condition were preventing manually-entered items to repopulate. 

I removed this condition; Instead, to ensure that only server with a valid `hostName` populates and entered in cookies _when_ read from .shi file - I set the condition [elsewhere](https://github.com/Moonshine-IDE/Moonshine-IDE/commit/f0ed559f27b524690db6b0b8e096b0b456cd9dbf#diff-7e94bf51f3cca2e6b1a3867641e0ab4e708f3d081263a6019176cd428633ec7bR161).

To test this properly, remove the `http://restapi...com:8080` marked broken server from Vagrant Instances, and re-start Moonshine; You won't see this appear again. However, if it once saved in cookie - it'll keep coming in the list.

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24370870/031fbd46-6d22-4005-b878-289718f737a1)

**Comment by piotrzarzycki21** [comment]
@rat-moonshine I'm not sure if your removal of my condition won't break #1242 - please try on your machine option from menu: **Project** -> **Deploy to Vagrant server** - you need to have active server - your **Launch url** should automatically adjust according to the server's parameters:

<img width="634" alt="Screenshot 2024-02-22 at 08 47 04" src="https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24554795/4470a2a2-8284-4aaa-89ca-2d85c2b93109">

Let me know if it's working for you.

**Comment by rat-moonshine** [comment]
The condition was a check during repopulation from saved-cookies. 

The condition anyway was populating faulty instance (i.e. instance without `hostName`) and preventing to populate manually-entered items - so the condition was wrong. A more proper condition however is now added elsewhere to test `hostName` availability before adding into cookie, as mentioned. Removal of previous condition not suppose to create any new trouble.

I don't have better setup to test your concern, but a test with whatever I have now, seems to works accordingly:

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24370870/7947d799-aa07-48ce-bacb-031a335d8d5b)

I see Bamboo Windows build was failing. It's restarted now. Please give some tests when it's available.

**Comment by JoelProminic** [comment]
I did some testing, and I confirmed the logic for #1242 still works on the imported Vagrant instances.   Manual entries use the default URLs.   The invalid entries I reported above should not trigger the #1242 logic anyway, since they would not be able to run and match the required capabilities for the action.

Since the update, I haven't been able to reproduce the case where my manual entries are cleared.   Some tests:

SHI Dev | Behavior
-- | --
No servers | Preserve manual list
Invalid server | Preserve manual list, no new entry
Valid server | Server added to list, manual entries preserved
